src/main/java/org/elasticsearch/indices/flush/SyncedFlushService.java

```diff
@@ -252,41 +315,24 @@ void sendSyncRequests(final String syncId, final List<ShardRouting> shards, Clus
                 results.put(shard, new SyncedFlushResponse("unknown node"));
-                contDownAndSendResponseIfDone(syncId, shards, shardId, listener, countDown, results);
+                contDownAndSendResponseIfDone(syncId, shards, shardId, totalShards, listener, countDown, results);
                 continue;
             }
             final Engine.CommitId expectedCommitId = expectedCommitIds.get(shard.currentNodeId());
             if (expectedCommitId == null) {
                 logger.trace("{} can't resolve expected commit id for node {}, skipping for sync id [{}]. shard routing {}", shardId, shard.currentNodeId(), syncId, shard);
                 results.put(shard, new SyncedFlushResponse("no commit id from pre-sync flush"));
-                contDownAndSendResponseIfDone(syncId, shards, shardId, listener, countDown, results);
+                contDownAndSendResponseIfDone(syncId, shards, shardId, totalShards, listener, countDown, results);
                 continue;
             }
             assert node.version().onOrAfter(Version.V_1_6_0) : " node with version " + node.version() + " can not have a sync commit ID";
             logger.trace("{} sending synced flush request to {}. sync id [{}].", shardId, shard, syncId);
-            final ActionListener<SyncedFlushResponse> currentShardListener = new ActionListener<SyncedFlushResponse>() {
-                @Override
-                public void onResponse(SyncedFlushResponse response) {
-                    SyncedFlushResponse existing = results.put(shard, response);
-                    assert existing == null : "got two answers for node [" + node + "]";
-                    // count after the assert so we won't decrement twice in handleException
-                    contDownAndSendResponseIfDone(syncId, shards, shardId, listener, countDown, results);
-                }
-
-                @Override
-                public void onFailure(Throwable e) {
-                    logger.trace("{} error while performing synced flush on [{}], skipping", e, shardId, shard);
-                    results.put(shard, new SyncedFlushResponse(e.getMessage()));
-                    contDownAndSendResponseIfDone(syncId, shards, shardId, listener, countDown, results);
-                }
-            };
             if (state.nodes().localNode().id().equals(node.id())) {
-                execute(ThreadPool.Names.FLUSH, new Callable<SyncedFlushResponse>() {
-                    @Override
-                    public SyncedFlushResponse call() throws Exception {
-                        return performSyncedFlush(new SyncedFlushRequest(shard.shardId(), syncId, expectedCommitId));
-                    }
-                }, currentShardListener);
+                SyncedFlushResponse existing = results.put(shard, performSyncedFlush(new SyncedFlushRequest(shard.shardId(), syncId, expectedCommitId)));
```

Comment:
I thought it was not needed anymore but think now it blocks and might not even be in the right thread pool...will add again. thanks also @spinscale